### PR TITLE
util: add PROXY protocol generation functions

### PR DESCRIPTION
**Pull request by wez470**
Description: Add utility functions for generating v1 and v2 PROXY protocol headers. This is the first step to adding Envoy PROXY proto generation support. These functions can also be used in other places that would also like to generate the header.
Risk Level: Low
Testing: Unit
Docs Changes: None
Release Notes: None
Related to https://github.com/envoyproxy/envoy/issues/1031

@alyssawilk

### CODE REVIEW

File: source/extensions/common/proxy_protocol/proxy_protocol_header.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+#include "envoy/network/address.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace ProxyProtocol {
+
+// See https://github.com/haproxy/haproxy/blob/master/doc/proxy-protocol.txt for definitions
+
+constexpr char PROXY_PROTO_V1_SIGNATURE[] = "PROXY ";
```
Comment (wez470) [inline source/extensions/common/proxy_protocol/proxy_protocol_header.h]: These constants overlap with ones created by the proxy proto listener filter. I opted to not refactor the listener filter to use these common ones for now to keep this PR simple. I can do the refactor now if requested however.

**Comment by wez470** [comment]
I saw the "envoy-linux (bazel clang_tidy)" build fail because it failed to fetch a repo. Thought I could re-run that with the retest thing but I guess not?
Screenshot_20200326_172512

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+#include "envoy/network/address.h"
+
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace ProxyProtocol {
+
+void generate_v1_header(const std::string& src_addr, const std::string& dst_addr, uint32_t src_port,
```
Comment (alyssawilk) [inline source/extensions/common/proxy_protocol/proxy_protocol_header.cc]: per style guide let's swap these from snake case towards generateV1Header etc.

File: source/extensions/common/proxy_protocol/proxy_protocol_header.h (hunk 1 above)
Comment (alyssawilk) [reply source/extensions/common/proxy_protocol/proxy_protocol_header.h]: How about a TODO so we don't lose track?

**Comment by wez470** [comment]
Thanks for the feedback! I'll move the regression tests over to this branch 🙂.

**Comment by alyssawilk** [comment]
LGTM pending CI cleanup :-)

**Comment by alyssawilk** [comment]
@mattklein123 or @zuercher @snowp any of you up for second pass?

File: source/extensions/common/proxy_protocol/proxy_protocol_header.cc [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace ProxyProtocol {
+
+void generateV1Header(const std::string& src_addr, const std::string& dst_addr, uint32_t src_port,
+                      uint32_t dst_port, Network::Address::IpVersion ip_version,
+                      Buffer::Instance& out) {
+  std::ostringstream stream;
```
Comment (mattklein123) [inline source/extensions/common/proxy_protocol/proxy_protocol_header.cc:18]: Sorry late drive by: my (possibly out of date) understanding is that string streams are pretty slow. We probably would want to move this to an envoy buffer or similar in a later perf pass.

**Pull request merged**